### ISSUE TITLE
[Bug]: Load balancer target not set for control-planes only cluster

### ISSUE THREAD

**Issue by mattanja** [kube-hetzner/terraform-hcloud-kube-hetzner#667]
### Description

With 3 control plane nodes and these settings, the targets of the load balancer are empty. I've tried with lb_hostname = var.cluster_domain set and unset.

`node.kubernetes.io/exclude-from-external-load-balancers=false` is set on every node (correctly) after !662 but still the targets are not set automatically.

### Kube.tf file

```terraform
locals {
  hcloud_token = var.hcloud_token
}

module "kube-hetzner" {
  providers = {
    hcloud = hcloud
  }
  hcloud_token = var.hcloud_token != "" ? var.hcloud_token : local.hcloud_token
  source = "kube-hetzner/kube-hetzner/hcloud"

  ssh_public_key = file("~/.ssh/id_rsa_hcloud.pub")
  ssh_private_key = file("~/.ssh/id_rsa_hcloud")
  network_region = "eu-central"

  control_plane_nodepools = [
    {
      name        = "node-nbg1",
      server_type = "cpx31",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 2,
      backups     = true
    },
    {
      name        = "node-fsn1",
      server_type = "cpx31",
      location    = "fsn1",
      labels      = [],
      taints      = [],
      count       = 1,
      backups     = true
    }
  ]

  agent_nodepools = [
    {
      name        = "agent-nbg1",
      server_type = "cpx31",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 0,
      # backups = true
    },
  ]

  load_balancer_type     = "lb11"
  load_balancer_location = "nbg1"
  base_domain = var.cluster_domain

  enable_longhorn = true
  disable_hetzner_csi = true

  allow_scheduling_on_control_plane = true
  cluster_name = "cluster01"

  use_control_plane_lb = false

  # Maybe this is the problem?
  # lb_hostname = var.cluster_domain
}

provider "hcloud" {
  token = var.hcloud_token != "" ? var.hcloud_token : local.hcloud_token
}

terraform {
  required_version = ">= 1.3.3"
  required_providers {
    hcloud = {
      source  = "hetznercloud/hcloud"
      version = ">= 1.35.2"
    }
  }
}

output "kubeconfig" {
  value     = module.kube-hetzner.kubeconfig
  sensitive = true
}

variable "hcloud_token" {
  sensitive = true
  default   = ""
}

variable "cluster_domain" {
  description = "TLD for the cluster"
  type        = string
  sensitive   = false
  default     = ""
}
```

### Screenshots

![image](https://user-images.githubusercontent.com/654416/226767099-44e694cf-3e35-46b0-a0bd-2a3e772377c6.png)

### Platform

Mac

**Comment by mysticaltech** [comment]
@mattanja I believe it may be working perfectly. The way to verify is just to issue `kubectl get svc -A` and see if the LB is pointing to the correct IPs, including those of the control plane.

Don't rely on the Hetzner LB health checks, those are meaningless unless you have an HTTP service responding to the LB checks, which you are free to implement. There is also an annotation that can be added to either Traefik or Nginx for that (search this project).

**Comment by mattanja** [comment]
@mysticaltech thanks for the fast reply. I'm not sure we're talking about the same thing. The issue is that the targets are not set and even removed every few hours if I set them manually. (Not talking about the health checks but the targets list is empty, no server is assigned.)

The get svc output looks fine (I guess):
`traefik           traefik                       LoadBalancer   10.43.XXX.XX    XXX.XXX.XXX.XXX,2a01:XXX:XXX:XXX::1   80:30182/TCP,443:30548/TCP   8h`

But still, the target servers of the Hetzner LoadBalancer keep getting removed and are not set up after initial setup.

**Comment by mysticaltech** [comment]
@mattanja Never even visit the UI, that's not built for Kubernetes, the Hetzner CCM manages the Kubernetes side, you can observe it's logs or describe it!

If the svc IPs are good, then all good 😊

**Comment by mattanja** [comment]
@mysticaltech no, still misunderstanding I think. Targets are not set! If I visit URLs pointing to the IP of the load-balancer, the request will never hit the cluster. It will end up on the load-balancer with no target to be routed to, resulting in error curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to url-pointing-to-load-balancer:443

**Comment by mattanja** [comment]
Providing some more info: in order to access the cluster via URLs I can manually set the targets, but they will be removed by the cluster setup (probably by the `hcloud-cloud-controller-manager`).

![image](https://user-images.githubusercontent.com/654416/226895691-36de4fa5-3eec-4d1e-9afb-c0c58b655e2e.png)

**Comment by mysticaltech** [comment]
@mattanja Please, while working with a LB deployed by Kubernetes via the Hetzner CCM, you should never ever touch anything manually.

When you create an ingress definition pointing to a service, Kubernetes and the Hetzner CCM will make sure that the targets are pointed to your service. NO MANUAL INTERVENTION should be done, otherwise it messes everything up, it's is an LB owned and requested by the the Hetzner CCM, humans should not touch it.

You can modify it via your ingress definitions, and/or the Traefik/Nginx ingress controller helm file (to change LB annotations for instance). But by default it works and if you see your Control plane nodes IPs when you do kubectl get svc -A, then all is as you wanted, which is pods get scheduled on the control plane and it is allowed as targets to the LB.

Also, I advise you to create a clean Hetzner project for your cluster, so you can see it better, avoiding mix-ups.

**Comment by mysticaltech** [comment]
Again, to really understand what is happening kubectl log --follow the Hetzer CCM pods.

**Comment by mysticaltech** [comment]
@mattanja If I am still misunderstanding, please provide your ingress definition with test pods and/or deployment definitions so that I can test myself.

**Comment by mattanja** [comment]
The manual modification was only to validate the issue. Ingress works fine after manually setting the Load-Balancer targets. I am aware that this should be done automatically, that's exactly the point. I delete and create completly new Hetzner cloud projects for testing the setup, there is no mix-up.

After the setup with the provided kube.tf there is essentially only two things that I'll do: Setup the clusterissuer and create a hello-world deployment and ingress.

I have now provided my testing setup here: https://github.com/mattanja/kube-hetzner-testing

**Comment by mattanja** [comment]
From the log of hcloud-cloud-controller-manager:
"There are no available nodes for LoadBalancer"


```
I0322 15:24:16.810012       1 load_balancers.go:108] "ensure Load Balancer" op="hcloud/loadBalancers.EnsureLoadBalancer" service="traefik" nodes=[]
I0322 15:24:16.810982       1 event.go:294] "Event occurred" object="traefik/traefik" fieldPath="" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
I0322 15:24:16.811003       1 event.go:294] "Event occurred" object="traefik/traefik" fieldPath="" kind="Service" apiVersion="v1" type="Warning" reason="UnAvailableLoadBalancer" message="There are no available nodes for LoadBalancer"
```

**Comment by mysticaltech** [comment]
@mattanja Well done and thanks for clarifying! 👍🚀

**Comment by mysticaltech** [comment]
@mattanja Your fix has been released in v1.10.7, thanks again for standing your ground and getting to the bottom of this. That's an important bug fix! 🙏
